### PR TITLE
conjur provider

### PR DESCRIPTION
**Pull request by psolarcz**
This PR addresses: https://github.com/external-secrets/external-secrets/issues/588 and add Cyberark Conjur Provider

### CODE REVIEW

**Comment by fbruandrade** [comment]
That's a great feature.... Im looking for a solution like this, do we have any updates?

**Review by infamousjoeg — CHANGES_REQUESTED** [review]
What is provided looks good to me. I would also ask prior to merge that relevant documentation be updated to reflect support for CyberArk Conjur Secrets Manager. (For example, README.md)

Also, would like to see an example of this in a manifest.  My fear is that a Conjur Host ID API Key will be added to the manifest in plain-text that could be accidentally committed to a public source code repository and someone could gain access to all secrets using that.

File: apis/externalsecrets/v1beta1/secretstore_conjur_types.go
Comment (moolen) [inline apis/externalsecrets/v1beta1/secretstore_conjur_types.go:22]: Conjur supports [different types of authentication](https://docs.conjur.org/Latest/en/Content/Developer/Conjur_API_Authenticate.htm) mechanisms, could we structure the CRD along the lines of the different means of authentication?

```yaml
# ...
conjur:
  auth:
    jwt:
      # server-hostname, service-id, account, host-id
    apiKey:
      # account, login, url
```

File: pkg/provider/conjur/provider.go
Comment (moolen) [inline pkg/provider/conjur/provider.go:101]: this does not allow using nested json structures or non-string values like float, int, array, null etc.

We may want to factor out this part here: https://github.com/external-secrets/external-secrets/blob/main/pkg/provider/vault/vault.go#L697-L719

**Review by moolen — COMMENTED** [review]
Hey! Sorry, this slipped through, apologies for the late response :disappointed: I've added a couple of comments. Don't hesitate to mention individuals or `@external-secrets/maintainers` and ask for a review :pray:!

**Comment by moolen** [comment]
Hey @psolarcz just wanted to check in and ask if you want to carry on with this PR?

**Comment by infamousjoeg** [comment]
> Hey @psolarcz just wanted to check in and ask if you want to carry on with this PR?

If they remain unresponsive, I'll follow through with resolving the review comments.

**Comment by psolarcz** [comment]
@infamousjoeg
Please continue with this PR, I don't use Conjur anymore so it's no longer priority for me.

**Comment by moolen** [comment]
@infamousjoeg do you want to open a new PR for this or work on his branch?
Just asking, i'd close this if you're opening a new one.

**Comment by infamousjoeg** [comment]
I'm going to continue working this one. This is on tap to be looked into within the next week.

**Comment by szh** [comment]
@infamousjoeg I just found out about this operator via another attendee of CloudNativeSecurityCon and was looking to see if it supported Conjur yet. Looks like you and @psolarcz beat me to it! Let me know if you want me to help out with this effort.

**Comment by moolen** [comment]
Hey @psolarcz @infamousjoeg are you working on this one, do you need any help?

**Comment by infamousjoeg** [comment]
@moolen I was able to deploy the external-secrets operator into my Kubernetes lab on Friday of last week.  I was also able to configure and successfully test Conjur as a secretStore using the Host/API authentication method.  I am beginning to review the requested changes here now that it is up and running, but unfortunately haven't had time yet this week to dedicate to it.  Regardless, this PR is being actively worked on, so it is not stale.

**Comment by gorkemozlu** [comment]
Having Conjur would enormously helpful, I see that PR is still active, @infamousjoeg did you had sometime to work this week ?

File: pkg/provider/conjur/provider.go
Comment (infamousjoeg) [reply pkg/provider/conjur/provider.go:101]: All values stored in Conjur are returned as string values. In my testing, this did not negatively effect the operator's ability to provide the secret variable from Conjur to Kubernetes Secrets.  Was there another reason this was suggested?

File: apis/externalsecrets/v1beta1/secretstore_conjur_types.go
Comment (infamousjoeg) [reply apis/externalsecrets/v1beta1/secretstore_conjur_types.go:22]: While Conjur does support authn-jwt as an additional method of authentication, we would like more time to develop that piece out through our Community & Integrations team.  We do, however, have customers who are okay with using the secretStore as provided and would prefer to release this as-is with a new version including authn-jwt to come at a later time.

Would this be possible?

cc: @szh

**Comment by infamousjoeg** [comment]
I have submitted a PR to @psolarcz fork that adds support for `serviceCertificate` for instances of Conjur that are using self-signed SSL certificates.  While, this is not recommended for Production scenarios, there will be many POCs that use this method where this will be a use case tested.

**Comment by gorkemozlu** [comment]
thanks a lot @infamousjoeg !

@psolarcz could you please add that PR into this PR ?

**Comment by psolarcz** [comment]
@infamousjoeg , @gorkemozlu  done. Sorry for the delay, I've missed 1st notification.

**Comment by gorkemozlu** [comment]
Hi @knelasevero , it looks like you're the reviewer, is that correct ?

**Comment by moolen** [comment]
@knelasevero or me, whoever gets to it first. Though i gotta review #1752 before this one :factory_worker:

**Comment by knelasevero** [comment]
Tested this quickly today, seems all fine. Also had a look over the code, looks good.

Can we squash and add sign-off by's?

Also there is a need to rebase and fix conflicts

File: pkg/provider/conjur/provider.go
Comment (moolen) [reply pkg/provider/conjur/provider.go:107]: The values are stored as string, right but when we unmarshal them it can break:

Works: simple string key/value pairs:

```json
{
  "foo": "bar"
}
```

This breaks:

```json
{
  "foo": 1234
}
```

And we can observe the error `unable to unmarshal secret BotApp/secretVar: json: cannot unmarshal number into Go value of type string`.

That raises the question: should the provider only support string/string key/value pairs?

If you want to support more you should take a look at [the vault provider implementation](https://github.com/external-secrets/external-secrets/blob/main/pkg/provider/vault/vault.go#L676-L704) which supports all data types and nested json as well. Though that can be totally added at a later stage.

File: apis/externalsecrets/v1beta1/secretstore_conjur_types.go
Comment (moolen) [inline apis/externalsecrets/v1beta1/secretstore_conjur_types.go:23]: Could you please add comments about  each field: what it is and where to get it.

The fields `ServiceURL`, `ServiceUser`, `ServiceAPIKey`, `ServiceAccount` are mandatory, hence `omitempty` can be removed. In fact, they can be a simple `string` instead of `*string`, that helps with `unexpected nil pointer dereference` errors in the future.

File: pkg/provider/conjur/provider.go
Comment (moolen) [inline pkg/provider/conjur/provider.go:54]: The `cfg.ServiceCertificate` can be `nil` and can throw a `nil pointer dereference` error.
It's probably best to make it a `string` instead of `*string`, as it seems the SDK handles that case well (empty string => no TLS).

File: apis/externalsecrets/v1beta1/secretstore_conjur_types.go
Comment (moolen) [inline apis/externalsecrets/v1beta1/secretstore_conjur_types.go:21]: The API key is a sensitive value and should be sourced from a secret resource. This is a pattern that is implemented throughout all the providers.

We have a [esmeta.SecretStoreSelector](https://github.com/external-secrets/external-secrets/blob/d605a5b450abb8072090ef115396f8b618aec2ec/apis/meta/v1/types.go#L19-L30) type that should be used for this case.

Note: `Namespace` must only be taken into account when using a ClusterSecretStore ([see here for example](https://github.com/external-secrets/external-secrets/blob/6de9a138db833cbe03da57682fff34117e94f175/pkg/provider/gcp/secretmanager/auth.go#L58-L65)). When a ClusterSecretStore is used then the `namespace` is mandatory.

A namespaced `SecretStore` must lookup the secret from the same namespace as the SecretStore resource itself. The `namespace` must not be set in this case.

Suggestion:
```yaml
apiVersion: external-secrets.io/v1beta1
kind: ClusterSecretStore
metadata:
  name: conjur-example
spec:
  provider:
    conjur:
      serviceUrl: "http://localhost:8080"
      serviceUser: "Dave@BotApp"
      serviceApiKey:
        secretRef:
          name: "conjur-creds"
          namespace: "external-secrets"
          key: "apikey"
```

File: pkg/provider/conjur/provider.go
Comment (moolen) [inline pkg/provider/conjur/provider.go:136]: The nil check is missing: `if conjurSpec.ServiceURL != nil && *conjurSpec.ServiceURL == ""`
Otherwise you have a nil pointer dereference panic and the controller dies.

I would suggest switching to a `string` value instead of using `*string` for all the fields.

File: apis/externalsecrets/v1beta1/secretstore_conjur_types.go
Comment (moolen) [reply apis/externalsecrets/v1beta1/secretstore_conjur_types.go]: Regarding the SecretStore API structure: I think it works for now but once we want to add a new authentication method it gets quite messy because all the fields are on the same level. I'm not saying all authentication methods must be implemented right away, but we should design the structure in a way that allows us to easily add another one.

This is not only for the sake of maintainability, but also a UX concern. From a design perspective i see two areas:
1. connection/server related settings: A `url` to configure and TLS-related settings (note: other providers use `caBundle` and `caProvider` [see here for example](https://external-secrets.io/v0.8.1/api/clustersecretstore/)). Potentially also timeouts.
2. How to authenticate: (apiKey today, maybe more in the future)

Suggestion:
```yaml
# ...
    conjur:
      server:
        url: "https://example.tech"
        caBundle: "..."
        caProvider: {} # ...
      auth:
        apiKey:
          secretRef:
            name: "conjur-creds"
            key: "apikey"
``` 

The `serviceCertificate` would be renamed to `caBundle` with type `[]byte` which contains the CA cert. `caProvider` can be added at a later stage to pull the ca cert from a Secret/ConfigMap.

WDYT?

**Review by moolen — COMMENTED** [review]
Hey, i'm glad we had a chance to meet at Kubecon :wave:. I found some time to review the PR and did a testdrive. I followed the `conjur-quickstart` repo which worked very nicely :+1:.

Overall it looks good to me, i would suggest restructuring the SecretStore CRD API. I left some comments, see below.

File: apis/externalsecrets/v1beta1/secretstore_conjur_types.go
Comment (knelasevero) [reply apis/externalsecrets/v1beta1/secretstore_conjur_types.go:21]: This is so true, can't believe I missed this one 😓 big +1

(to the other comments as well)

Comment (infamousjoeg) [reply apis/externalsecrets/v1beta1/secretstore_conjur_types.go:21]: I was wondering how we could do this! Thank you for this information! I'm looking into it now. ❤️

**Comment by munkyxtc** [comment]
@moolen is there anything pending on this one or just queued up?

**Comment by moolen** [comment]
Hey @munkyxtc, a couple of things need to be addressed before we can merge, See comments above.

**Comment by knelasevero** [comment]
@infamousjoeg would you be able to address the comments from Moritz? Let us know if this PR should have someone else take over. (Red Hat now also has interest in getting this one merged :D)

**Comment by szh** [comment]
FYI @infamousjoeg is now on parental leave

**Comment by infamousjoeg** [comment]
> @infamousjoeg would you be able to address the comments from Moritz? Let us know if this PR should have someone else take over. (Red Hat now also has interest in getting this one merged :D)

@munkyxtc is taking over for me as I am on parental leave now until end of August.  Also, @szh is working with Red Hat from our side.  Both should be able to guide this PR to merge.

**Comment by davidh-cyberark** [comment]
> > @infamousjoeg would you be able to address the comments from Moritz? Let us know if this PR should have someone else take over. (Red Hat now also has interest in getting this one merged :D)
> 
> @munkyxtc is taking over for me as I am on parental leave now until end of August. Also, @szh is working with Red Hat from our side. Both should be able to guide this PR to merge.

Hey, @infamousjoeg & @munkyxtc , I'm actively working on updating the yaml structure proposed by @moolen .  I'm working to have something ready for review next week.

What is left for review on this PR besides the YAML update?

Regards,
David

**Comment by moolen** [comment]
Closing this in favor of #2412 which just got merged and will be released with `0.9` which will hit until the end of this week.

Thanks to everyone involved in this, we're happy to get feedback on the first iteration on this new provider :sparkles:

**Pull request closed without merging**
